ios/FluentUI/Bottom Sheet/CommandingItem.swift

```diff
@@ -0,0 +1,102 @@
+///
+/// `CommandingItem` defines the high level properties and behavior of a command. Its visual representation is determined by
+/// the `BottomCommandingController`.
+@objc(MSFCommandingItem)
+open class CommandingItem: NSObject {
+
+    /// A closure that's called when the command is triggered
+    @objc open var action: (CommandingItem) -> Void
+
+    /// The title of the command item.
+    @objc open var title: String {
+        didSet {
+            if title != oldValue {
+                delegate?.commandingItem(self, didChangeTitleFrom: oldValue)
+            }
+        }
+    }
+
+    /// A `UIImage` to be displayed with the command.
+    @objc open var image: UIImage {
+        didSet {
+            if image != oldValue {
+                delegate?.commandingItem(self, didChangeImageFrom: oldValue)
+            }
+        }
+    }
+
+    /// A `UIImage` used when the command is represented as a button in selected state.
+    @objc open var selectedImage: UIImage? {
+        didSet {
+            if selectedImage != oldValue {
+                delegate?.commandingItem(self, didChangeSelectedImageFrom: oldValue)
+            }
+        }
+    }
+
+    /// Indicates whether the command is currently on.
+    ///
+    /// When `commandType` is set to `.toggle`, the item toggles this automatically before calling `action`.
+    @objc open var isOn: Bool {
+        didSet {
+            if isOn != oldValue {
+                delegate?.commandingItem(self, didChangeOnFrom: oldValue)
+            }
+        }
+    }
+
+    /// Indicates whether the command is enabled.
+    @objc open var isEnabled: Bool {
+        didSet {
+            if isEnabled != oldValue {
+                delegate?.commandingItem(self, didChangeEnabledFrom: oldValue)
+            }
+        }
+    }
+
+    /// Determines the behavior of this command when its triggered.
+    @objc open var commandType: CommandType {
+        didSet {
+            if commandType != oldValue {
+                delegate?.commandingItem(self, didChangeCommandTypeFrom: oldValue)
+            }
+        }
+    }
+
+    @objc public init(title: String, image: UIImage, action: @escaping (CommandingItem) -> Void, selectedImage: UIImage? = nil, isSelected: Bool = false, isEnabled: Bool = true, commandType: CommandType = .simple) {
+        self.title = title
+        self.action = action
+        self.image = image
+        self.selectedImage = selectedImage
+        self.isOn = isSelected
+        self.isEnabled = isEnabled
+        self.commandType = commandType
+    }
+
+    @objc public enum CommandType: Int {
+        /// Calls `action` on tap without modifying the item.
+        case simple
+
+        /// Toggles `isOn` on tap before calling `action`.
+        case toggle
+    }
+
+    weak var delegate: CommandingItemDelegate?
+}
+
+protocol CommandingItemDelegate: class {
+    func commandingItem(_ item: CommandingItem, didChangeTitleFrom oldValue: String)
```

Comment:
why do we need to tell the old value via the protocol? if it didChange we should just follow the changed value?

Reply:
For me this was more of a "why not". The internal property observers have the old value available to them in didSet, and it can be used to make things more efficient, like no-oping when the value hasn't changed. The delegate is kind of an external observer in this case, so I didn't see why not give them the old value as well. They can still get the current one directly from the item, just like in local didSets.

Reply:
but anytime protocol didSomthingFoo is only called when it actually did change something. this is different from property setting don't you think?

Reply:
Makes sense, I changed it to pass the new value. I wasn't using the old one anyways and if we need it in the future it's easy to add.